src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs

```diff
@@ -0,0 +1,300 @@
+
+            using var session = CreateSessionFactory().OpenSession();
+            using var tx = session.BeginTransaction();
+
+            var copyStagedTimeoutsBackToTimeoutsHqlQuery = session.CreateQuery(@"
+INSERT INTO TimeoutEntity (
+    Id,
+    Destination,
+    SagaId,
+    State,
+    Endpoint,
+    Time,
+    Headers
+)
+SELECT
+    STE.Id,
+    STE.Destination,
+    STE.SagaId,
+    STE.State,
+    STE.Endpoint,
+    STE.Time,
+    STE.Headers
+FROM StagedTimeoutEntity STE
+WHERE STE.BatchState <> :BatchCompletedStatus;
+");
+            copyStagedTimeoutsBackToTimeoutsHqlQuery.SetParameter("BatchCompletedStatus", BatchState.Completed);
+            await copyStagedTimeoutsBackToTimeoutsHqlQuery.ExecuteUpdateAsync();
+
+            var deleteStagedTimeoutsHqlQuery = session.CreateQuery(@"
+DELETE StagedTimeoutEntity STE
+WHERE STE.BatchState <> :BatchCompletedStatus;");
+            deleteStagedTimeoutsHqlQuery.SetParameter("BatchCompletedStatus", BatchState.Completed);
+            await deleteStagedTimeoutsHqlQuery.ExecuteUpdateAsync();
+
+            var markMigrationEntityAsAbortedHqlQuery = session.CreateQuery(@"UPDATE MigrationsEntity
+SET
+    CompletedAt= :CompletedAt,
+    Status = :AbortedStatus
+WHERE
+    MigrationRunId = :MigrationRunId;");
+            markMigrationEntityAsAbortedHqlQuery.SetParameter("CompletedAt", DateTime.UtcNow);
+            markMigrationEntityAsAbortedHqlQuery.SetParameter("AbortedStatus", MigrationStatus.Aborted);
+            markMigrationEntityAsAbortedHqlQuery.SetParameter("MigrationRunId", ((NHibernateToolState)migrationRun).MigrationRunId);
+            await markMigrationEntityAsAbortedHqlQuery.ExecuteUpdateAsync();
+
+            await tx.CommitAsync();
+        }
+
+        public async Task<IReadOnlyList<EndpointInfo>> ListEndpoints(DateTime cutOffTime)
+        {
+            using var session = CreateSessionFactory().OpenStatelessSession();
+
+            TimeoutEntity timeoutAlias = null;
+            var timeouts = await session.QueryOver(() => timeoutAlias)
+                .Where(timeout => timeout.Time >= cutOffTime)
+                .SelectList(list => list
+                    .SelectGroup(p => p.Endpoint)
+                    .SelectMax(p => p.Time)
+                    .SelectMin(p => p.Time)
+                    .SelectCount(p => p.Id)
+                    ).ListAsync<object[]>();
+
+            return timeouts.Select(x => new EndpointInfo
+                {
+                    EndpointName = x[0].ToString(),
+                    LongestTimeout = (DateTime)x[1],
+                    ShortestTimeout = (DateTime)x[2],
+                    NrOfTimeouts = (int)x[3],
+                    Destinations = session.QueryOver<TimeoutEntity>()
+                                            .Where(timeout => timeout.Endpoint == x[0].ToString())
+                                            .SelectList(t => t.SelectGroup(x => x.Destination))
```

Comment:
I was wondering if we could use Nhibernates way of projecting into non-tracked entities here and if that would make it more readable. Not against this approach by the way and maybe it is not worth the extra effort. Just a thought